RSpec/ScatteredLet unexpected behavior on let(:x, &proc)

I've come to a weird use of `let` today:

```ruby
shared_examples 'validate event_type' do |args|
  ...
  let(:expected_event_type_to_use, &args[:event_type_to_use])
  ...
end

...
it_behaves_like 'validate event_type', event_type_to_use: -> { event_type_sub.id }
...
```

Not going to discussion if this was a good use or not, it just works. RuboCop is flagging it wrongly though: it seems that it does not flag it as `let` and therefore all the `let`s below it are flagged by `RSpec/ScatteredLet`:
```text
159:7  warning  rubocop:C  RSpec/ScatteredLet: Group all let/let! blocks in the example group together.
 164:7  warning  rubocop:C  RSpec/ScatteredLet: Group all let/let! blocks in the example group together.
 165:7  warning  rubocop:C  RSpec/ScatteredLet: Group all let/let! blocks in the example group together.
 180:7  warning  rubocop:C  RSpec/ScatteredLet: Group all let/let! blocks in the example group together.
 185:7  warning  rubocop:C  RSpec/ScatteredLet: Group all let/let! blocks in the example group together.
 190:7  warning  rubocop:C  RSpec/ScatteredLet: Group all let/let! blocks in the example group together.
 195:7  warning  rubocop:C  RSpec/ScatteredLet: Group all let/let! blocks in the example group together.
```

Version info:
```
⋊> ruby -v
ruby 2.6.2p47 (2019-03-13 revision 67232) [x86_64-darwin17]
⋊> rails -v
Rails 5.2.3
⋊> bundle exec rubocop -v
0.64.0
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,7 @@
 * Add new `Capybara/VisibilityMatcher` cop. ([@aried3r][])
 * Ignore String constants by `RSpec/Describe`. ([@AlexWayfer][])
 * Drop support for ruby 2.3. ([@bquorning][])
+* Fix multiple cops to detect `let` with proc argument. ([@tejasbubane][])
 
 ## 1.38.1 (2020-02-15)
 
@@ -497,3 +498,4 @@ Compatibility release so users can upgrade RuboCop to 0.51.0. No new features.
 [@eitoball]: https://github.com/eitoball
 [@aried3r]: https://github.com/aried3r
 [@AlexWayfer]: https://github.com/AlexWayfer
+[@tejasbubane]: https://github.com/tejasbubane

--- a/lib/rubocop/rspec/language.rb
+++ b/lib/rubocop/rspec/language.rb
@@ -28,6 +28,14 @@ module RuboCop
           "(block #{send_pattern} ...)"
         end
 
+        def block_pass_pattern
+          "(send #{RSPEC} #{node_pattern_union} _ block_pass)"
+        end
+
+        def block_or_block_pass_pattern
+          "{#{block_pattern} #{block_pass_pattern}}"
+        end
+
         def send_pattern
           "(send #{RSPEC} #{node_pattern_union} ...)"
         end

--- a/lib/rubocop/rspec/language/node_pattern.rb
+++ b/lib/rubocop/rspec/language/node_pattern.rb
@@ -17,7 +17,7 @@ module RuboCop
 
         def_node_matcher :hook?, Hooks::ALL.block_pattern
 
-        def_node_matcher :let?, Helpers::ALL.block_pattern
+        def_node_matcher :let?, Helpers::ALL.block_or_block_pass_pattern
 
         def_node_matcher :subject?, Subject::ALL.block_pattern
       end

--- a/spec/rubocop/cop/rspec/align_left_let_brace_spec.rb
+++ b/spec/rubocop/cop/rspec/align_left_let_brace_spec.rb
@@ -45,6 +45,12 @@ RSpec.describe RuboCop::Cop::RSpec::AlignLeftLetBrace do
   end
   # rubocop:enable RSpec/ExampleLength
 
+  it 'does not register offense for let with proc argument' do
+    expect_no_offenses(<<-RUBY)
+      let(:user, &args[:build_user])
+    RUBY
+  end
+
   it 'works with empty file' do
     expect_no_offenses('')
   end

--- a/spec/rubocop/cop/rspec/align_right_let_brace_spec.rb
+++ b/spec/rubocop/cop/rspec/align_right_let_brace_spec.rb
@@ -45,6 +45,12 @@ RSpec.describe RuboCop::Cop::RSpec::AlignRightLetBrace do
   end
   # rubocop:enable RSpec/ExampleLength
 
+  it 'does not register offense for let with proc argument' do
+    expect_no_offenses(<<-RUBY)
+      let(:user, &args[:build_user])
+    RUBY
+  end
+
   it 'works with empty file' do
     expect_no_offenses('')
   end

--- a/spec/rubocop/cop/rspec/empty_line_after_final_let_spec.rb
+++ b/spec/rubocop/cop/rspec/empty_line_after_final_let_spec.rb
@@ -47,6 +47,26 @@ RSpec.describe RuboCop::Cop::RSpec::EmptyLineAfterFinalLet do
     RUBY
   end
 
+  it 'checks for empty line after let with proc argument' do
+    expect_offense(<<-RUBY)
+      RSpec.describe User do
+        let(:a) { a }
+        let(:user, &args[:build_user])
+        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Add an empty line after the last `let` block.
+        it { expect(a).to eq(b) }
+      end
+    RUBY
+
+    expect_correction(<<-RUBY)
+      RSpec.describe User do
+        let(:a) { a }
+        let(:user, &args[:build_user])
+
+        it { expect(a).to eq(b) }
+      end
+    RUBY
+  end
+
   it 'approves empty line after let' do
     expect_no_offenses(<<-RUBY)
     RSpec.describe User do

--- a/spec/rubocop/cop/rspec/leading_subject_spec.rb
+++ b/spec/rubocop/cop/rspec/leading_subject_spec.rb
@@ -41,6 +41,25 @@ RSpec.describe RuboCop::Cop::RSpec::LeadingSubject do
     RUBY
   end
 
+  it 'checks subject below let with proc argument' do
+    expect_offense(<<-RUBY)
+      RSpec.describe User do
+        let(:user, &args[:build_user])
+
+        subject { described_class.new }
+        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Declare `subject` above any other `let` declarations.
+      end
+    RUBY
+
+    expect_correction(<<-RUBY)
+      RSpec.describe User do
+        subject { described_class.new }
+        let(:user, &args[:build_user])
+
+      end
+    RUBY
+  end
+
   it 'approves of subject above let' do
     expect_no_offenses(<<-RUBY)
       RSpec.describe User do

--- a/spec/rubocop/cop/rspec/let_before_examples_spec.rb
+++ b/spec/rubocop/cop/rspec/let_before_examples_spec.rb
@@ -62,6 +62,25 @@ RSpec.describe RuboCop::Cop::RSpec::LetBeforeExamples do
     RUBY
   end
 
+  it 'flags `let` with proc argument' do
+    expect_offense(<<-RUBY)
+      RSpec.describe User do
+        include_examples('should be after let')
+
+        let(:user, &args[:build_user])
+        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Move `let` before the examples in the group.
+      end
+    RUBY
+
+    expect_correction(<<-RUBY)
+      RSpec.describe User do
+        let(:user, &args[:build_user])
+        include_examples('should be after let')
+
+      end
+    RUBY
+  end
+
   it 'does not flag `let` before the examples' do
     expect_no_offenses(<<-RUBY)
       RSpec.describe User do

--- a/spec/rubocop/cop/rspec/scattered_let_spec.rb
+++ b/spec/rubocop/cop/rspec/scattered_let_spec.rb
@@ -25,4 +25,15 @@ RSpec.describe RuboCop::Cop::RSpec::ScatteredLet do
       end
     RUBY
   end
+
+  it 'flags `let` with proc argument' do
+    expect_offense(<<-RUBY)
+      RSpec.describe User do
+        let(:a) { a }
+        subject { User }
+        let(:user, &args[:build_user])
+        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Group all let/let! blocks in the example group together.
+      end
+    RUBY
+  end
 end

